sys/linux/dev_kvm.txt

```diff
@@ -275,18 +275,24 @@ syzos_api_smccc {
 	arg_params	array[int64, 5]
 }
 
+syzos_api_irq_setup {
+	nr_cpus	const[1, int32]
```

Comment:
Why const 1?

Reply:
Here the situation is even more complicated. Right now the VCPU doesn't know its own ID, because it is only passed to KVM_CREATE_VCPU and cannot be retrieved later. Upon calling syz_kvm_setup_cpu() every CPU is identical, so we need to somehow pass that VCPU ID to syz_kvm_setup_cpu().

Could it be possible to declare a syscall's input parameter as a resource?
Something along the lines of:

```
r0 = openat$kvm(0, &AUTO='/dev/kvm\x00', 0x0, 0x0)
r1 = ioctl$KVM_CREATE_VM(r0, AUTO, 0x0)
# The value of r3 is set to the input parameter of KVM_CREATE_VCPU that takes values from 0 to 2
r2 = ioctl$KVM_CREATE_VCPU(r1, AUTO, r3=0x0)
# Passing r3 somewhere to tell the CPU its ID
syz_kvm_setup_cpu$arm64(r1, r2, ... r3)
```

Reply:
Let's probably pass 0:4 here as well and see how it goes.

Reply:
> Is there an ioctl that returns vcpu id? It can be used in syz_kvm_setup_cpu if it exists

No, there is none. Normally userspace doesn't need it, because the caller of KVM_CREATE_VCPU assings the id.

> Btw, do we need vcpu_id or vcpu_idx?

Never heard of vcpu_idx, and https://elixir.bootlin.com/linux/v6.11-rc2/source/arch/arm64/kvm/vgic/vgic.c#L433 is using vcpu_id. Having to access vcpu_idx would be even worse, as now we'll need to maintain a mapping between the two.

> Id is possible to get as resource using an "identity" pseudo-syscall that just returns the inputs argument.
But the problem is that if it's not tied to vcpu resource, then there is no guarantee they will be matched. I.e. if we create vcpu1/id1 and vcpu2/id2, the fuzzer can pass vcpu1/id2 to syz_kvm_setup_cpu. At this point it's not much different from using an int with small range (int32[0:2]).

Fair enough, that's what I'll do then.